src/basic_types.hpp

```diff
@@ -62,6 +55,31 @@ enum class BoundaryType : int {
 // E3 = z-aligned edge at (i - 1/2, j - 1/2, k)
 // NN = node at (i - 1/2, j - 1/2, k - 1/2)
 //
+// Some select topological elements around cell (i,j,k) with o corresponding
+// to faces, x corresponding to edges, and + corresponding to nodes (the indices
+// denote the array index of each element):
+// clang-format off
+//
+//                      E1(i,j+1,k+1)
+//              NN_+---------x---------+_NN(i+1,j+1,k+1)
+//     (i,j+1,k+1)/|  F3(i,j,k+1)     /|
+//               / |      |          / |
+//           E2_x  |      o         x__|_E2(i+1,j,k+1)
+//    (i,j,k+1)/   x         o     /   x___E3(i+1,j+1,k)
+//            /    |         |___ /____|_F2(i,j+1,k)
+//        NN_+---------x---------+_____|_NN(i+1,j,k+1)
+// (i,j,k+1) |  o  |  E1         |  o__|____F1(i+1,j,k)
```

Comment:
nice! this must have taken forever to create...

Reply:
It wasn't too bad, although I did start from an ASCII cube I found somewhere on the internet.